line_graph for multigraphs

fabien-vignes opened the pull request:
We generalize the `line_graph` method to graphs possibly with multiple edges and/or loops.
It is useful in order to compute the line-graph of a multigraph but also to generate the edge-colorings of a multigraph from the `all_graph_coloring` method.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

fabien-vignes committed:
line_graph for multigraphs

fabien-vignes committed:
blank line removed

dcoudert commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -351,17 +396,35 @@ def line_graph(g, labels=True):
+    if multiple:
+        labels = True
+
+    h = g.copy()
```
you can avoid this copy when `g` is simple. Consider the following
```sage
    multiple = g.has_multiple_edges()
    if multiple:
        labels = True  # this must be documented
        origlabels_dic = {(u, v, id): (u, v, label)
                          for id, (u, v, label) in enumerate(g.edge_iterator())}
        g = parent(g)([g, origlabels_dic.keys()], format='vertices_and_edges',
                      multiedges=True)
```
After that, you continue to work with `g` and not `h`.

dcoudert commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -263,21 +263,32 @@ def is_line_graph(g, certificate=False):
+      of multiple edges, the vertices of the line graph will be triples
+      ``(u,v,an integer)``.
+
+    - ``origlabels`` -- boolean (default: ``False``); wether edge labels should
```
The name of this parameter is not really nice. `return_labels` ? 

`wether` -> `whether`

dcoudert reviewed (changes requested):
I have other possible changes in mind, to get a faster code and to avoid calls to `hash`, but we can do that a future PR. Let us first finalize this one.

A more important question is if we should directly label the vertices of the line-graph with integers instead of tuples.

dcoudert commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -311,7 +322,7 @@ def line_graph(g, labels=True):
          (1, 3, None),
          (2, 3, None)]
-        sage: h.am()                                                                    # needs sage.modules
+        sage: h.am()        # needs sage.modules
```
please don't modify the indentation of these tags. The large number of spaces is on purpose to avoid polluting the documentation.

dcoudert reviewed (commented):
Feel free to integrate all the changes I proposed.

Also, you may add the definition you use for the line graph of a multigraph. As we discussed, we have 2 cases: connect 2 vertices if the corresponding edges are incident in $g$, or connect 2 edges if the corresponding edges have exactly one extremity in common in $g$. You use the first definition. We may add the second later.

fabien-vignes committed:
Modifications suggested by David Coudert

Copy of the input graph g only if g.has_multiple_edges().
New copy is called g as the input.

fchapoton commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -371,18 +371,18 @@ def line_graph(g, labels=True, origlabels=False):
         (1, 2, 4): 'c'}
         sage: g = DiGraph([(0,1),(0,1),(1,2)],multiedges=True)
-        sage: g.line_graph().edges()
+        sage: line_graph(g).edges()
```
wrong move here, c'est une methode..

dcoudert commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -387,13 +446,13 @@ def line_graph(g, labels=True):
 
     G.add_vertices(elist)
-
+        
```
Remove the trailing white spaces

dcoudert commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -387,13 +446,13 @@ def line_graph(g, labels=True):
 
         # Add the edge to the list, according to hashes, as previously
-        for e in g.edge_iterator(v, labels=labels):
+        for e in g.edge_iterator(v, labels=labels): # iterates over the edges incident to v
```
please add 2 spaces before `#`

fabien-vignes committed:
line_graph(g) -> g.line_graph() dans Examples

fabien-vignes committed:
Modifications to pass the tests (hopefully)

- Removed white spaces in blanklines (in the Examples)
- Add import command for parent

fabien-vignes committed:
Moved cimport parent at the top of the file

fchapoton commented on src/sage/graphs/line_graph.pyx:
```diff
@@ -1,4 +1,6 @@
 # cython: binding=True
+
+from sage.structure.element cimport parent
```
mettre l'import plutot ligne 131, juste apres la documentation globale

fabien-vignes committed:
Modification of examples and new location for import of parent

In case return_labels == True, the returned dictionary is different from
a previous version. An example output had to be modified accordingly.

fabien-vignes committed:
Added a second blank line below an import statement

dcoudert reviewed (approved):
LGTM.

Thank you for this improvement.
